### PR TITLE
SONAR-10126 add path of project basedir and last revision id to scanner report

### PR DESCRIPTION
**Pull request by janos-ss**

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/scm/ScmProvider.java [hunk 1]
```diff
@@ -64,4 +64,19 @@ public BlameCommand blameCommand() {
...
+  public String projectBasedir() {
+    return ".";
+  }
+
+  /**
+   * The corresponding revision ID of the pull request,
+   * for example the SHA1 of the last commit in a Git branch.
+   */
+  public String revisionId() {
+    return "";
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/scm/ScmProvider.java]: MINOR Remove this method and declare a constant for this value. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3400)

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/scm/ScmProvider.java [hunk 2]
```diff
@@ -64,4 +64,19 @@ public BlameCommand blameCommand() {
   public Set<Path> branchChangedFiles(String targetBranchName, Path rootBaseDir) {
     return null;
   }
+
+  /**
+  * The relative path to the project from SCM root, typically ".".
+  */
+  public String projectBasedir() {
+    return ".";
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/scm/ScmProvider.java]: MINOR Remove this method and declare a constant for this value. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3400)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/report/MetadataPublisher.java [hunk 3]
```diff
@@ -47,15 +50,20 @@
   private final ScannerPluginRepository pluginRepository;
   private final BranchConfiguration branchConfiguration;
 
+  @Nullable
+  private final ScmProvider scmProvider;
+
   public MetadataPublisher(ProjectAnalysisInfo projectAnalysisInfo, InputModuleHierarchy moduleHierarchy, Configuration settings,
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/report/MetadataPublisher.java]: MAJOR Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00107)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/scm/ScmProvider.java (hunk 2 above)
Comment (henryju) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/scm/ScmProvider.java]: I don't like this default value. "." doesn't have a meaning in itself. If we are talking about returning the relative path of sqProjectBaseDir comparing to scmRoot, we could return Path.get("").

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/scm/ScmProvider.java [hunk 4]
```diff
@@ -64,4 +64,19 @@ public BlameCommand blameCommand() {
...
+  }
+
+  /**
+   * The corresponding revision ID of the pull request,
```
Comment (henryju) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/scm/ScmProvider.java]: Why "of the pull request"? I would say: "The current revision ID of the analyzed code"

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/scm/ScmProvider.java (hunk 1 above)
Comment (henryju) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/scm/ScmProvider.java]: I would prefer to return UnsupportedOperationException (and catch it in the calling code).

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/scm/ScmProvider.java (hunk 2 above)
Comment (henryju) [reply sonar-plugin-api/src/main/java/org/sonar/api/batch/scm/ScmProvider.java]: Another option is to throw UnsupportedOperationException and catch it in the calling code, to use the appropriate default value.

File: sonar-scanner-protocol/src/main/protobuf/scanner_report.proto [hunk 5]
```diff
@@ -43,6 +43,10 @@ message Metadata {
   BranchType branch_type = 10;
   string merge_branch_name = 11;
 
+  // pull request support, see PullRequestConfiguration
+  string project_basedir = 12;
+  string revision_id = 13;
```
Comment (henryju) [inline sonar-scanner-protocol/src/main/protobuf/scanner_report.proto]: Having scm in the name would be clearer. scm_revision_id?

File: sonar-scanner-protocol/src/main/protobuf/scanner_report.proto [hunk 6]
```diff
@@ -43,6 +43,10 @@ message Metadata {
   BranchType branch_type = 10;
   string merge_branch_name = 11;
 
+  // pull request support, see PullRequestConfiguration
+  string project_basedir = 12;
```
Comment (henryju) [inline sonar-scanner-protocol/src/main/protobuf/scanner_report.proto]: relative_path_from_scm_root ?

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonar-scanner-engine/src/test/java/org/sonar/scanner/mediumtest/ScannerMediumTester.java [hunk 7]
```diff
@@ -300,6 +303,7 @@ public TaskResult execute() {
           tester.projectRefProvider,
           tester.activeRules,
           tester.serverIssues,
+          mock(ScmConfiguration.class),
```
Comment (henryju) [inline sonar-scanner-engine/src/test/java/org/sonar/scanner/mediumtest/ScannerMediumTester.java]: I'm surprised we have to inject a mock for medium tests. The purpose is to test real components as far as possible.

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/report/MetadataPublisher.java [hunk 8]
```diff
@@ -47,15 +57,25 @@
   private final ScannerPluginRepository pluginRepository;
   private final BranchConfiguration branchConfiguration;
 
+  @Nullable
+  private final ScmConfiguration scmConfiguration;
+
   public MetadataPublisher(ProjectAnalysisInfo projectAnalysisInfo, InputModuleHierarchy moduleHierarchy, Configuration settings,
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/report/MetadataPublisher.java]: MAJOR Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00107)

File: plugins/sonar-xoo-plugin/src/main/java/org/sonar/xoo/scm/XooScmProvider.java [hunk 9]
```diff
@@ -47,4 +48,16 @@ public BlameCommand blameCommand() {
...
 
+  @Override
+  public Path relativePathFromScmRoot(Path path) {
+    return findScmRoot(path).relativize(path);
+  }
+
+  private Path findScmRoot(Path path) {
+    Path work = path;
+    while (!work.resolve(".xoo").toFile().isDirectory()) {
+      work = work.getParent();
```
Comment (henryju) [inline plugins/sonar-xoo-plugin/src/main/java/org/sonar/xoo/scm/XooScmProvider.java]: possible NPE if no .xoo is found.

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/TestInputFileBuilder.java [hunk 10]
```diff
@@ -108,13 +110,22 @@ public static int nextBatchId() {
...
+    this.projectBaseDir = normalize(projectBaseDir);
+    return this;
+  }
+
   public TestInputFileBuilder setModuleBaseDir(Path moduleBaseDir) {
+    this.moduleBaseDir = normalize(moduleBaseDir);
+    return this;
+  }
+
+  private Path normalize(Path path) {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/TestInputFileBuilder.java]: MAJOR Make "normalize" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/report/ComponentsPublisher.java [hunk 11]
```diff
@@ -187,7 +193,26 @@ private String getPath(InputComponent component) {
...
+  @CheckForNull
+  private String getProjectRelativePath(DefaultInputComponent component) {
+    if (component instanceof InputFile) {
+      DefaultInputFile inputFile = (DefaultInputFile) component;
+      return inputFile.getProjectRelativePath();
+    }
+
+    Path projectBaseDir = moduleHierarchy.root().getBaseDir();
+    if (component instanceof InputDir) {
+      InputDir inputDir = (InputDir) component;
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/report/ComponentsPublisher.java]: MINOR Remove this use of "InputDir"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
MINOR Remove this use of "InputDir"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/report/ComponentsPublisher.java [hunk 12]
```diff
@@ -187,7 +193,26 @@ private String getPath(InputComponent component) {
...
+
+  @CheckForNull
+  private String getProjectRelativePath(DefaultInputComponent component) {
+    if (component instanceof InputFile) {
+      DefaultInputFile inputFile = (DefaultInputFile) component;
+      return inputFile.getProjectRelativePath();
+    }
+
+    Path projectBaseDir = moduleHierarchy.root().getBaseDir();
+    if (component instanceof InputDir) {
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/report/ComponentsPublisher.java]: MINOR Remove this use of "InputDir"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**